File: src/services/thongKeService.js
import axios from "axios";
import SERVER_URL from "../configs/server.config";

// Import ExcelJS trực tiếp
import ExcelJS from "exceljs";

// Dynamic import cho jsPDF để tránh lỗi babel runtime
let jsPDF = null;
const loadJsPDF = async () => {
  if (!jsPDF) {
    const jsPDFModule = await import("jspdf");
    jsPDF = jsPDFModule.default || jsPDFModule.jsPDF;
  }
  return jsPDF;
};

const API_URL = `${SERVER_URL}/zmen`;

// Helper function để format currency
const formatCurrency = (value) => {
  if (value == null) return "0 VNĐ";
  if (typeof value === "number") {
    return new Intl.NumberFormat("vi-VN", {
      style: "currency",
      currency: "VND",
    }).format(value);
  }
  return `${value} VNĐ`;
};

// Helper function để check API có khả dụng không
const isApiAvailable = async () => {
  try {
    await axios.get(`${API_URL}/test`, {
      timeout: 3000,
    });
    return true;
  } catch (error) {
    console.warn("API không khả dụng:", error.message);
    return false;
  }
};

const thongKeService = {
  // ==================== SẢN PHẨM BÁN CHẠY ====================
  getSanPhamBanChayTheoNgay: async (ngay) => {
    try {
      const response = await axios.get(`${API_URL}/san-pham/ngay`, {
        params: { ngay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching san pham ban chay theo ngay:", error);
      return [];
    }
  },

  getSanPhamBanChayTheoThang: async (thang, nam) => {
    try {
      const response = await axios.get(`${API_URL}/san-pham/thang`, {
        params: { thang, nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching san pham ban chay theo thang:", error);
      return [];
    }
  },

  getSanPhamBanChayTheoNam: async (nam) => {
    try {
      const response = await axios.get(`${API_URL}/san-pham/nam`, {
        params: { nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching san pham ban chay theo nam:", error);
      return [];
    }
  },

  getSanPhamBanChayKhoangNgay: async (tuNgay, denNgay) => {
    try {
      const response = await axios.get(`${API_URL}/san-pham/khoang-ngay`, {
        params: { tuNgay, denNgay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching san pham ban chay khoang ngay:", error);
      return [];
    }
  },

  // ==================== DOANH THU ====================
  getDoanhThuTheoNgay: async (ngay) => {
    try {
      const response = await axios.get(`${API_URL}/doanh-thu/ngay`, {
        params: { ngay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching doanh thu theo ngay:", error);
      return [];
    }
  },

  getDoanhThuTheoThang: async (thang, nam) => {
    try {
      const response = await axios.get(`${API_URL}/doanh-thu/thang`, {
        params: { thang, nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching doanh thu theo thang:", error);
      return [];
    }
  },

  getDoanhThuTheoNam: async (nam) => {
    try {
      const response = await axios.get(`${API_URL}/doanh-thu/nam`, {
        params: { nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching doanh thu theo nam:", error);
      return [];
    }
  },

  getDoanhThuKhoangNgay: async (tuNgay, denNgay) => {
    try {
      const response = await axios.get(`${API_URL}/doanh-thu/khoang-ngay`, {
        params: { tuNgay, denNgay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching doanh thu khoang ngay:", error);
      return [];
    }
  },

  // ==================== HIỆU SUẤT NHÂN VIÊN ====================
  getHieuSuatNVTheoNgay: async (ngay) => {
    try {
      const response = await axios.get(`${API_URL}/nhan-vien/ngay`, {
        params: { ngay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching hieu suat nhan vien theo ngay:", error);
      return [];
    }
  },

  getHieuSuatNVTheoThang: async (thang, nam) => {
    try {
      const response = await axios.get(`${API_URL}/nhan-vien/thang`, {
        params: { thang, nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching hieu suat nhan vien theo thang:", error);
      return [];
    }
  },

  getHieuSuatNVTheoNam: async (nam) => {
    try {
      const response = await axios.get(`${API_URL}/nhan-vien/nam`, {
        params: { nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching hieu suat nhan vien theo nam:", error);
      return [];
    }
  },

  getHieuSuatNVKhoangNgay: async (tuNgay, denNgay) => {
    try {
      if (await isApiAvailable()) {
        const response = await axios.get(`${API_URL}/nhan-vien/khoang-ngay`, {
          params: { tuNgay, denNgay },
        });
        return response.data;
      } else {
        return null; // Return null if API is not available
      }
    } catch (error) {
      console.error("Error fetching hieu suat nhan vien khoang ngay:", error);
      return null; // Return null if API is not available
    }
  },

  // ==================== KHÁCH HÀNG CHI TIÊU ====================
  getKhachHangChiTieuTheoNgay: async (ngay) => {
    try {
      const response = await axios.get(`${API_URL}/khach-hang/ngay`, {
        params: { ngay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching khach hang chi tieu theo ngay:", error);
      return [];
    }
  },

  getKhachHangChiTieuTheoThang: async (thang, nam) => {
    try {
      const response = await axios.get(`${API_URL}/khach-hang/thang`, {
        params: { thang, nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching khach hang chi tieu theo thang:", error);
      return [];
    }
  },

  getKhachHangChiTieuTheoNam: async (nam) => {
    try {
      const response = await axios.get(`${API_URL}/khach-hang/nam`, {
        params: { nam },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching khach hang chi tieu theo nam:", error);
      return [];
    }
  },

  getKhachHangChiTieuKhoangNgay: async (tuNgay, denNgay) => {
    try {
      const response = await axios.get(`${API_URL}/khach-hang/khoang-ngay`, {
        params: { tuNgay, denNgay },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching khach hang chi tieu khoang ngay:", error);
      return [];
    }
  },

  // ==================== THỐNG KÊ TỔNG QUAN ====================
  getThongKeTongQuan: async () => {
    try {
      const response = await axios.get(`${API_URL}/tong-quan`);
      return response.data;
    } catch (error) {
      console.error("Error fetching thong ke tong quan:", error);
      return null;
    }
  },

  // ==================== BIỂU ĐỒ DOANH THU ====================
  getBieuDoDoanhThu: async (filterType, year, month) => {
    try {
      const response = await axios.get(`${API_URL}/doanh-thu/bieu-do`, {
        params: { filterType, year, month },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching bieu do doanh thu:", error);
      return [];
    }
  },

  // ==================== DOANH THU THEO DANH MỤC ====================
  getDoanhThuTheoDanhMuc: async (year, month) => {
    try {
      const response = await axios.get(`${API_URL}/danh-muc/doanh-thu`, {
        params: { year, month },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching doanh thu theo danh muc:", error);
      return [];
    }
  },

  // ==================== THỐNG KÊ THANH TOÁN ====================
  getThongKeThanhToan: async (nam, thang) => {
    try {
      const response = await axios.get(`${API_URL}/thanh-toan/thong-ke`, {
        params: { nam, thang },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching thong ke thanh toan:", error);
      return [];
    }
  },

  // ==================== DANH SÁCH ĐƠN HÀNG ====================
  getDonHangList: async (
    filterType,
    selectedDate,
    selectedMonth,
    selectedYear,
    dateFrom,
    dateTo
  ) => {
    try {
      const response = await axios.get(`${API_URL}/don-hang/list`, {
        params: {
          filterType,
          selectedDate,
          selectedMonth,
          selectedYear,
          dateFrom,
          dateTo,
        },
      });
      return response.data;
    } catch (error) {
      console.error("Error fetching don hang list:", error);
      return [];
    }
  },

  // ==================== EXPORT EXCEL ====================
  exportSanPhamBanChayExcel: async (ngay) => {
    try {
      const response = await axios.get(
        `${API_URL}/san-pham/ngay/export-excel`,
        {
          params: { ngay },
          responseType: "blob",
        }
      );
      return response.data;
    } catch (error) {
      console.error("Error exporting san pham ban chay excel:", error);
      throw error;
    }
  },

  exportDoanhThuExcel: async (loai, ngay, thang, nam) => {
    try {
      const response = await axios.get(`${API_URL}/doanh-thu/export-excel`, {
        params: { loai, ngay, thang, nam },
        responseType: "blob",
      });
      return response.data;
    } catch (error) {
      console.error("Error exporting doanh thu excel:", error);
      throw error;
    }
  },

  exportTongQuanExcel: async () => {
    try {
      const response = await axios.get(`${API_URL}/tong-quan/export-excel`, {
        responseType: "blob",
      });
      return response.data;
    } catch (error) {
      console.error("Error exporting tong quan excel:", error);
      throw error;
    }
  },

  // ==================== EXPORT THỐNG KÊ MỚI ====================
  // Export Excel với dữ liệu đầy đủ
  exportThongKeExcel: async (data) => {
    try {
      const response = await axios.post(
        `${API_URL}/export/thong-ke-excel`,
        data,
        {
          responseType: "blob",
        }
      );
      return response.data;
    } catch (error) {
      console.error("Error exporting Excel:", error);
      return null;
    }
  },

  // Export PDF với dữ liệu đầy đủ
  exportThongKePDF: async (data) => {
    try {
      const response = await axios.post(
        `${API_URL}/export/thong-ke-pdf`,
        data,
        {
          responseType: "blob",
        }
      );
      return response.data;
    } catch (error) {
      console.error("Error exporting PDF:", error);
      return null;
    }
  },
};

export default thongKeService;

File: Website-b-n-qu-n-o-nam-ZMEN/ZMEN/src/main/java/com/example/ZMEN/service/ThongKeService.java
package com.example.ZMEN.service;

import com.example.ZMEN.dto.ThongKeDto.DoanhThuDto;
import com.example.ZMEN.dto.ThongKeDto.HieuSuatNVDto;
import com.example.ZMEN.dto.ThongKeDto.KhachHangChiTieuDto;
import com.example.ZMEN.dto.ThongKeDto.SanPhamBanChayDto;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public interface ThongKeService {
    // ==================== SẢN PHẨM BÁN CHẠY ====================
    List<SanPhamBanChayDto> thongKeTheoNgay(LocalDate ngay);
    List<SanPhamBanChayDto> thongKeTheoThang(int thang, int nam);
    List<SanPhamBanChayDto> thongKeTheoNam(int nam);
    List<SanPhamBanChayDto> thongKeTheoKhoangNgay(LocalDate tuNgay, LocalDate denNgay);

    // ==================== DOANH THU ====================
    DoanhThuDto tinhDoanhThuTheoNgay(LocalDate ngay);
    DoanhThuDto tinhDoanhThuTheoThang(int thang, int nam);
    DoanhThuDto tinhDoanhThuTheoNam(int nam);
    DoanhThuDto tinhDoanhThuTheoKhoangNgay(LocalDate tuNgay, LocalDate denNgay);
    
    // Các method mới cho doanh thu
    DoanhThuDto getDoanhThuTheoNgay(LocalDate ngay);
    DoanhThuDto getDoanhThuTheoThang(int thang, int nam);
    DoanhThuDto getDoanhThuTheoNam(int nam);
    DoanhThuDto getDoanhThuKhoangNgay(LocalDate tuNgay, LocalDate denNgay);

    // ==================== HIỆU SUẤT NHÂN VIÊN ====================
    List<HieuSuatNVDto> hieuSuatNhanVienTheoNgay(LocalDate ngay);
    List<HieuSuatNVDto> hieuSuatNhanVienTheoThang(int thang, int nam);
    List<HieuSuatNVDto> hieuSuatNhanVienTheoNam(int nam);
    List<HieuSuatNVDto> hieuSuatNhanVienTheoKhoangNgay(LocalDate tuNgay, LocalDate denNgay);
    
    // Các method mới cho hiệu suất nhân viên
    List<HieuSuatNVDto> getHieuSuatNVTheoNgay(LocalDate ngay);
    List<HieuSuatNVDto> getHieuSuatNVTheoThang(int thang, int nam);
    List<HieuSuatNVDto> getHieuSuatNVTheoNam(int nam);
    List<HieuSuatNVDto> getHieuSuatNVKhoangNgay(LocalDate tuNgay, LocalDate denNgay);

    // ==================== KHÁCH HÀNG CHI TIÊU ====================
    List<KhachHangChiTieuDto> khachHangChiTieuTheoNgay(LocalDate ngay);
    List<KhachHangChiTieuDto> khachHangChiTieuTheoThang(int thang, int nam);
    List<KhachHangChiTieuDto> khachHangChiTieuTheoNam(int nam);
    List<KhachHangChiTieuDto> khachHangChiTieuTheoKhoangNgay(LocalDate batDau, LocalDate ketThuc);
    
    // Các method mới cho khách hàng chi tiêu
    List<KhachHangChiTieuDto> getKhachHangChiTieuTheoNgay(LocalDate ngay);
    List<KhachHangChiTieuDto> getKhachHangChiTieuTheoThang(int thang, int nam);
    List<KhachHangChiTieuDto> getKhachHangChiTieuTheoNam(int nam);
    List<KhachHangChiTieuDto> getKhachHangChiTieuKhoangNgay(LocalDate tuNgay, LocalDate denNgay);

    // ==================== DANH SÁCH ĐƠN HÀNG ====================
    List<Map<String, Object>> getDonHangList(String filterType, LocalDate selectedDate, Integer selectedMonth, Integer selectedYear, LocalDate dateFrom, LocalDate dateTo);

    // ==================== THỐNG KÊ TỔNG QUAN ====================
    Long getTongDonHang();
    Long getTongKhachHang();
    Long getTongSanPham();
    Double getDoanhThuHomNay();
    Double getDoanhThuThangNay();
    Double getDoanhThuNamNay();
    Double getTangTruongThang();
    Double getTangTruongNam();
    Integer getKhachHangMoi();
    Double getTyLeHuy();
    List<Map<String, Object>> getTopSanPhamBanChayThang();
    List<Map<String, Object>> getTopKhachHangThang();
    
    // Method tổng hợp cho thống kê tổng quan
    Map<String, Object> getThongKeTongQuan();

    // ==================== BIỂU ĐỒ VÀ PHÂN TÍCH ====================
    List<Map<String, Object>> getBieuDoDoanhThu(String loai, Integer nam, Integer thang);
    List<Map<String, Object>> getDoanhThuTheoDanhMuc(Integer nam, Integer thang);
    List<Map<String, Object>> getThongKeThanhToan(Integer nam, Integer thang);
}
